src/storage/src/source/mod.rs

```diff
@@ -439,6 +447,12 @@ pub trait SourceReader {
         }
     }
 
+    /// Partitions that this reader knows about but is not consuming from. We
+    /// need these to compute a "global" source upper, when determining
+    /// completeness of a timestamp.
+    // WIP: Maybe find a better name?
+    fn unconsumed_partitions(&self) -> Vec<PartitionId>;
```

Comment:
I think `-> &[PartitionId]` would be nicer!

as for naming, I think `known_uncovered_partitions` might be better but its a hard one as its a very niche concept